marshmallow_dataclass/__init__.py

```diff
@@ -163,10 +172,46 @@ def dataclass(
     >>> Point.Schema().load({'x':0, 'y':0}) # This line can be statically type checked
     Point(x=0.0, y=0.0)
     """
-    # dataclass's typing doesn't expect it to be called as a function, so ignore type check
-    dc = dataclasses.dataclass(  # type: ignore
-        _cls, repr=repr, eq=eq, order=order, unsafe_hash=unsafe_hash, frozen=frozen
-    )
+    # Check python version for dataclass params only available for python >= 3.10.
+    # If python version is below 3.10 and any of these params are set to anything
+    # other than their default, raise ValueError
+    if sys.version_info >= (3, 10):
+        # dataclass's typing doesn't expect it to be called as a function, so ignore type check
+        dc = dataclasses.dataclass(  # type: ignore
+            _cls,
+            repr=repr,
+            eq=eq,
+            order=order,
+            unsafe_hash=unsafe_hash,
+            frozen=frozen,
+            match_args=match_args,
+            kw_only=kw_only,
+            slots=slots,
+        )
+    else:
+        args = {
+            "match_args": {
+                "error_message": "'match_args' argument is only available for python >= 3.10",
+                "default_value": True,
+            },
+            "kw_only": {
+                "error_message": "'kw_only' argument is only available for python >= 3.10",
+                "default_value": False,
+            },
+            "slots": {
+                "error_message": "'slots' argument is only available for python >= 3.10",
+                "default_value": False,
+            },
+        }
+
+        for arg, arg_info in args.items():
+            if locals()[arg] is not arg_info["default_value"]:
+                raise ValueError(arg_info["error_message"])
+
+        # dataclass's typing doesn't expect it to be called as a function, so ignore type check
+        dc = dataclasses.dataclass(  # type: ignore
+            _cls, repr=repr, eq=eq, order=order, unsafe_hash=unsafe_hash, frozen=frozen
+        )
```

Comment:
Why this ? Can't we just forward the arguments we got to the base dataclass implementation, and let it error if it wants to ?

Reply:
Agree.  For the implementation can we not just do
```py
def dataclass(
    __cls: Optional[Type[_U]] = None,
   *,
   base_schema: Optional[Type[marshmallow.Schema]] = None,
   cls_frame: Optional[types.FrameType] = None,
   **kwargs: bool,
 ) -> Union[Type[_U], Callable[[Type[_U]], Type[_U]]]:
    dc = dataclasses.dataclass(**kwargs)
    if cls_frame is None:
        # ... logic to find frame unchanged ...

    def decorate(cls: Type[_U]) -> Type[_U]:
        return add_schema(dc(cls), base_schema, cls_frame=cls_frame)

    if _cls is None:
        return decorate
    return decorate(cls)
```

Reply:
This breaks the type signature.